dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,443 @@
+type ResourceClaim struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec defines what to allocated and how to configure it.
+	Spec ResourceClaimSpec `json:"spec"`
+
+	// Status describes whether the claim is ready for use.
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
+	Status ResourceClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
 
-	// DeviceConfigs contains references to arbitrary vendor device configuration
-	// objects that will be attached to the device allocation.
+type ResourceClaimSpec struct {
+	// This field holds configuration for multiple potential drivers which
+	// could satisfy requests in this claim. The configuration applies to
+	// the entire claim. It is ignored while allocating the claim.
 	//
 	// +optional
 	// +listType=atomic
-	Configs []DeviceClassConfigReference `json:"configs,omitempty"`
-}
+	Config []ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,4,opt,name=config"`
 
-// Namespace scoped.
-type DeviceClaim struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// These constraints must be satisfied by the set of devices that get
+	// allocated for the claim.
+	Constraints []Constraint `json:"constraints"`
 
-	Spec   DeviceClaimSpec   `json:"spec,omitempty"`
-	Status DeviceClaimStatus `json:"status,omitempty"`
+	// Requests are individual requests for separate resources for the claim.
+	// An empty list is valid and means that the claim can always be allocated
+	// without needing anything. A class can be referenced to use the default
+	// requests from that class.
+	//
+	// +listType=atomic
+	Requests []Request `json:"requests,omitempty" protobuf:"bytes,5,name=requests"`
+
+	// Future extension, ignored by older schedulers. This is fine because scoring
+	// allows users to define a preference, without making it a hard requirement.
+	//
+	//
+	// Score *SomeScoringStruct
+
+	// Shareable indicates whether the allocated claim is meant to be shareable
+	// by multiple consumers at the same time.
+	// +optional
+	Shareable bool `json:"shareable,omitempty" protobuf:"bytes,3,opt,name=shareable"`
 }
 
-// DeviceClaimSpec details the requirements that devices chosen
-// to satisfy this claim must meet.
-type DeviceClaimSpec struct {
-	// MatchAttributes allows specifying a constraint that will apply
-	// across all of the claims. For example, if you specified "numa", then
-	// this overall claim could only be successfully fulfilled if all of
-	// the included claims could be fulfilled by pools with the same "numa"
-	// attribute value. If we simply set matchAttributes in each claim
-	// separately, then they could be consistent within claims, but
-	// inconsistent across claims. Therefore, we need this additional
-	// constraint.
+// Constraint must have one and only one field set.
```

Comment:
For posterity: This is a proper use of list-of-one-of.  We have potentially more than 1 criterion at the same time, and this will signal to back-rev components that there is something here they don't understand.